Add config file with path and classname options - Fixes #192

defenestrator opened the pull request:
Fixes #192 by adding config file for paths for Livewire views, javascript assets and class namespaces.

Currently, Livewire hardcodes the applications root namespace, and view path. This change allows for configuration of those values to be pulled from .env, to be set statically in the config file, or retain the existing defaults. Includes a third configuration value to set the js asset "base_url"

(Do over from yesterday when I had no business sitting at a keyboard! :honeybee:)

defenestrator committed:
Add config file with path and classname options - Fixes #192

defenestrator committed:
config file itself

calebporzio commented:
Thanks @defenestrator, where is the config file stub though? Am I missing something? I loved that you had proper Taylor Otwell comments.

calebporzio commented:
Oh, lol - nvm

defenestrator commented:
About fifty percent of the time I make a new directory I fail to `git add` it, lately - new workflows are new, and I'm rusty, thanks.

defenestrator committed:
Potentially sets $appUrl value twice,
but separate `if` statements preserve existing logic in the documentation
and allows config override via $options[]

defenestrator committed:
merge upstream

tillkruss commented:
Maybe this PR could also default to `Application::getNamespace()` in case the config file doesn't exist?

defenestrator commented:
> Maybe this PR could also default to `Application::getNamespace()` in case the config file doesn't exist?

The config file feels like the wrong place to do it, in terms of conventions, any preferences about that? New class or...I haven’t read enough of the source to know where to stick it yet. I will get back to this tomorrow.

tillkruss commented:
That wouldn’t actually work I believe. Sorry I should have been more specific. 

The `config()` helper accepts a default value as the 2nd parameter 👌🏻

defenestrator commented:
Of course it does, and somehow I have never used it, thanks @tillkruss

tillkruss commented on src/LivewireServiceProvider.php:
```diff
@@ -72,6 +74,10 @@ public function boot()
+
+        $this->publishes([
+            __DIR__.'/../config/livewire.php' => base_path('config/livewire.php'),
+        ], 'config');
```
This should be `livewire-config`, not just `config`.

tillkruss commented:
```
php artisan vendor:publish --tag=livewire-config
```

defenestrator committed:
Merge remote-tracking branch 'upstream/master'

defenestrator commented:
I have other improvements to make and some tests to add, but got caught up in other stuff yesterday.
I can throw those over the wall today.

defenestrator committed:
Defaults and namespace detection,
clean up just a bit.

defenestrator committed:
Remove space to make StyleCI happy

defenestrator commented:
This is approximately good to go. I do want to add some new tests, but in looking into that I learned some refactoring of the existing tests might be wise, and probably needs a separate issue and PR.

defenestrator committed:
fix view path name

defenestrator committed:
namespace clarity

calebporzio commented on config/livewire.php:
```diff
@@ -0,0 +1,43 @@
+    |
+    */
+
+    'base_url'  => env('LIVEWIRE_BASE_URL', '/'),
```
inconsistent casing between "view-path" and "base_url".

I'm going to just throw notes in here because I see you are already committing right now, otherwise, I wouldn't hold up a merge for tiny things like this. Thanks a ton for this PR.

defenestrator replied:
That's because of existing kebab casing convention in the view names, I don't like it either, but didn't want my scope to be too creepy. Will address

calebporzio commented on config/livewire.php:
```diff
@@ -0,0 +1,43 @@
+
+    'base_url'  => env('LIVEWIRE_BASE_URL', '/'),
+];
+/*  */
```
Extra empty comment here?

defenestrator replied:
Not sure where that came from, haha, I have cleanup to do. 
I'm finally starting to grok some of the source here, I like where you're going with this library.

calebporzio replied:
Awesome to hear @defenestrator!

calebporzio commented on src/Commands/LivewireMakeCommandParser.php:
```diff
@@ -52,8 +57,10 @@ public function classFile()
 
     public function classNamespace()
     {
+        $namespace = config('livewire.namespace', $this->getAppNamespace());
```
I'd maybe group all the items fetched from the config in the constructor.

defenestrator replied:
So funny I had it that way and then moved it into the method.

calebporzio commented on src/Component.php:
```diff
@@ -128,7 +128,7 @@ public function castDataToJavaScriptReadableTypes($data)
         array_walk($data, function ($value, $key) {
             throw_unless(
-                is_bool($value) || is_null($value) || is_array($value) || is_numeric($value) || is_string($value),
+                is_null($value) || is_array($value) || is_numeric($value) || is_string($value),
```
Is this a regression?

calebporzio commented on src/LivewireManager.php:
```diff
@@ -177,7 +177,15 @@ protected function javaScriptAssets($jsFileName, $options)
-        $appUrl = rtrim($options['base_url'] ?? '', '/');
+        $appUrl = '';
+
+        if (file_exists(config_path('livewire.php'))) {
```
rather than checking for the existence of the config file, maybe just use the aforementioned second param of the config helper for a default. It's more flexible that way (thinking things like testing).

defenestrator replied:
When I thought this through before it made sense to me to do it this way for some reason, will review.

calebporzio commented on src/LivewireServiceProvider.php:
```diff
@@ -32,6 +32,8 @@ public function register()
 
         $this->allowCertainExceptionsToBypassTheBladeViewHandler();
+
+        $this->mergeConfigFrom(__DIR__.'/../config/livewire.php', 'livewire');
```
Do we still need this if we use the second param defaults?

defenestrator replied:
Nope.

calebporzio commented on src/LivewireServiceProvider.php:
```diff
@@ -84,7 +90,7 @@ public function registerRoutes()
     public function registerViews()
     {
-        $this->loadViewsFrom(__DIR__.DIRECTORY_SEPARATOR.'views', 'livewire');
+        $this->loadViewsFrom(__DIR__.DIRECTORY_SEPARATOR.'views', config('livewire.view-path'));
```
second param default here? Not sure, just asking.

defenestrator replied:
about to commit it, yes.

calebporzio reviewed (changes requested):
Great stuff, couple notes for you. THANKS!

defenestrator committed:
cleanup and improvements per comments on PR

defenestrator committed:
StyleCI

defenestrator committed:
I'm a dingbat

defenestrator commented:
I'll get back to this in a bit @calebporzio life happened and now I'm making dumb mistakes - thanks for the feedback.

defenestrator commented:
Sorry, a tree fell on my house Sunday, I will get back and finish this...today? Soon as possible.

calebporzio commented:
Hey @defenestrator, that's crazy - sorry about your house. I'm actually heading out of town for a week, so I'll get back to this when I get back. Thanks!

calebporzio commented:
I want this in, so I'm going to just merge this and make some of the changes myself. Thanks a ton for the work on this @defenestrator!!

calebporzio committed:
Merge branch 'master' into master

defenestrator commented:
I feel awful about this PR. I had to replace the roof on my house myself
and then my girlfriend and I both came down with freaking Listeria! Almost
a month later and we’re barely feeling human again. Home ownership is an
odyssey. Cheers to the Livewire team, I hope to make real use of it soon.

On Wed, Sep 18, 2019 at 14:08 Caleb Porzio <notifications@github.com> wrote:

> Merged #235 <https://github.com/calebporzio/livewire/pull/235> into
> master.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/calebporzio/livewire/pull/235?email_source=notifications&email_token=AAEBIJWLPMZIJZJRCPWDG6DQKKDFVA5CNFSM4ITJJ752YY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGOTWSD7BY#event-2644787079>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEBIJTM4VM3337CIEBHMX3QKKDFVANCNFSM4ITJJ75Q>
> .
>
-- 
With my warm regard,

Jeremy Anderson
📱1.208.283.8654
- via mobile

lancepioch commented:
@all-contributors please add @defenestrator for code
